finetune/base_models/bert/featurizer.py

```diff
@@ -140,4 +145,279 @@ def bert_featurizer(
+        """
+        X_sub = tf.compat.v1.Print(
+            X_sub, ["X_sub", X_sub, tf.shape(X_sub)], summarize=-1
+        )
+        # To fit the interface of finetune we are going to compute the mask and type id at runtime.
+        delimiters = tf.cast(tf.equal(X_sub, encoder.delimiter_token), tf.int32)
+        # Appears to be 0 where you have tokens and 1 where you have padding
+        # [batch_size, sequence_length]
+        token_type_ids = tf.cumsum(delimiters, exclusive=True, axis=1)
+        seq_length = tf.shape(input=delimiters)[1]
+        eos_idx = tf.argmax(
+            input=tf.cast(delimiters, tf.float32)
+            * tf.expand_dims(
+                tf.range(tf.cast(seq_length, tf.float32), dtype=tf.float32), 0
+            ),
+            axis=1,
+        )
+        lengths = lengths_from_eos_idx(eos_idx=eos_idx, max_length=seq_length)
+        # Appears to be 1 where you have tokens and 0 where you have padding
+        # [batch_size, sequence_length]
+        mask = tf.sequence_mask(lengths, maxlen=seq_length, dtype=tf.float32)
+        bert = underlying_model(
+            config=bert_config,
+            is_training=train,
+            input_ids=X_sub,
+            input_mask=mask,
+            input_context=context,
+            token_type_ids=token_type_ids,
+            use_one_hot_embeddings=False,
+            scope=None,
+            use_pooler=config.bert_use_pooler,
+            use_token_type=config.bert_use_type_embed,
+            reading_order_decay_rate=reading_order_decay_rate,
+        )
+        return bert.get_pooled_output()
+
+    def chunk_aggregation(pooled_outputs):
+        """
+        Function for aggregating/pooling the pooled outputs from the BERT featurizer
+        across the chunk dimension
+
+        TODO Implement max, attention, lstm, and concat options
+        """
+        if config.chunk_pool_fn == "mean":
+            aggr_outputs = tf.reduce_mean(pooled_outputs, axis=0)
+        elif config.chunk_pool_fn == "attention":
+            aggr_outputs = attn(pooled_outputs)
+        # This is wrong
+        else:
+            aggr_outputs = tf.reduce_mean(pooled_outputs, axis=0)
+
+        return aggr_outputs
+
+    with tf.compat.v1.variable_scope("model/featurizer", reuse=reuse):
+        # Get pooled output across chunk dim using map_fn
+        # [num_chunks / batch_scaling, batch_size * batch_scaling, hidden_size]
+        # pooled_output = tf.map_fn(lambda inp: bert_wrapper(inp[0], inp[1], inp[2]), (X, mask, token_type_ids))
+        pooled_output = tf.map_fn(lambda inp: bert_wrapper(inp), X, dtype=tf.float32)
+        pooled_output = tf.compat.v1.Print(
+            pooled_output, ["pooled_output", tf.shape(pooled_output)], summarize=-1
+        )
+
+        # Reshape output to remove batch_scaling factor
+        # [num_chunks, batch_size, hidden_size]
+        pool_shape = tf.shape(input=pooled_output)
+        pooled_output = tf.reshape(
```

Comment:
I think you may need to reverse the transpose ([https://github.com/IndicoDataSolutions/finetune/pull/639/files#diff-792a67b685da819a4dba8219f199f61c2505680332dd26f7cc35814dfa043322R250](here)) before the reshape?

Reply:
Yup at least the attn op in the chunk aggregation expects batch dim to be 0.